### DISCUSSION TITLE
Intersection processing

### DISCUSSION THREAD

**Discussion by BruceBrown** [dabreegster/abstreet#429] [General]
Do intersections support right turn on red? From the intersection editor, I'd guess they don't, as there is no way to indicate "no right on red". I'm assuming the same would be true of left turn on red when all streets at the intersection are one-way. Any thoughts on adding support for it?

**Comment by dabreegster** [comment]
![screencast](https://user-images.githubusercontent.com/1664407/102724504-c4efb880-42c4-11eb-9b61-e02ae27c2a09.gif)
(Edited to use a GIF; the mp4 didn't work)

Both should be supported. The video shows a car turning right on red at the very beginning, then I edit the intersection to disable right turn on red, and move the red turns into their own stage.

The dotted lines in the editor are meant to show unprotected turns where you may have to yield to higher-priority traffic. Designing the UI of this editor has been quite difficult; any idea how to more clearly show right-turn-on-red there?

**Reply by BruceBrown** [reply]
nice. I missed that you could click on the lines. Let me think about the UI a bit.

I'm running the signal scenario and failing at E. Maddison and 12th, where I have a car and bike both wanting to move forward into the same lane and neither doing so, both stuck for 10 minutes.

**Comment by dabreegster** [comment]
> I'm running the signal scenario and failing at E. Maddison and 12th

This intersection is a mess:
![Screenshot from 2020-12-20 13-42-01](https://user-images.githubusercontent.com/1664407/102725125-329de380-42c9-11eb-86a6-606d4233d0cf.png)
E Union technically does come in at a "separate" intersection: https://www.openstreetmap.org/node/3447400930
But really these two should be merged and treated as one: https://github.com/dabreegster/abstreet/issues/114#issuecomment-743414012

In the short term, this is broken without any easy fix. If you want to dig into why neither vehicle is going, you can enable dev mode (**Ctrl+S**), then debug mode (**Ctrl+D**), then the blocked-by graph explorer (**b**)

**Reply by BruceBrown** [reply]
I got passed that mess --- just had to add dedicated crossing and an extra west-bound cycle. However now I have pedestrians causing jams at stop sign intersections. There's no way to change the intersection and no way to queue the pedestrians; so a long chain separated by a few seconds apart becomes a solid fence. I watched a jam at S Jackson and 16th form where southbound couldn't turn east and west bound couldn't proceed. All this, despite the fact the lanes were clear for blocks. I may upload a short video to capture the frustration. :)

**Reply by dabreegster** [reply]
> There's no way to change the intersection and no way to queue the pedestrians

This is a known issue; I don't think we have a bug filed with all of the details yet. There are a few problems:

1) We're not using any sidewalk or crosswalk info from OpenStreetMap, instead generating crossings at every possible intersection. This is super unrealistic.
2) Pedestrians always have priority over vehicles, and there's no way for the player to configure this.
3) A pedestrian (or vehicle taking an unprotected turn) will start their movement through an intersection if there's not a higher-priority vehicle currently turning. But they won't check for a vehicle that's a few seconds away from the intersection.

**Comment by BruceBrown** [comment]
On Sun, Dec 20, 2020 at 1:45 PM Dustin Carlino <notifications@github.com>
wrote:

> I'm running the signal scenario and failing at E. Maddison and 12th
>
> This intersection is a mess:
> [image: Screenshot from 2020-12-20 13-42-01]
> <https://user-images.githubusercontent.com/1664407/102725125-329de380-42c9-11eb-86a6-606d4233d0cf.png>
> E Union technically does come in at a "separate" intersection:
> https://www.openstreetmap.org/node/3447400930
> But really these two should be merged and treated as one: #114 (comment)
> <https://github.com/dabreegster/abstreet/issues/114#issuecomment-743414012>
>
> In the short term, this is broken without any easy fix. If you want to dig
> into why neither vehicle is going, you can enable dev mode (*Ctrl+S*),
> then debug mode (*Ctrl+D*), then the blocked-by graph explorer (*b*)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dabreegster/abstreet/discussions/429#discussioncomment-227849>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABVX7FR4LS5TH5N62OCQCTSVZV6HANCNFSM4VDMAWHA>
> .
>

Are cars well behaved in taking turns at a stop sign? Specifically, I'm
looking at Yesler Way and Terrace St. The car on Terrace Street would like
to turn W. However the cars on Yesler Way Stop and Go without letting the
Stopped Terrace St car advance. This seems like a bug in the algorithm.
[image: Screen Shot 2020-12-21 at 11.02.08 AM.png]

**Reply by dabreegster** [reply]
Your image didn't seem to come through on Github, so I'm not 100% sure what issue you're seeing. It might be the 3rd problem I mentioned in the other thread, or it could be an issue with lane selection

**Reply by BruceBrown** [reply]
I'm thinking about adding a new "adaptive" with the following parameters: Variable(skip,min,delay,max), where skip is true if the cycle can be skipped if no demand, min is the minimum cycle time, delay is the elapsed zero demand to the end the cycle, e.g, 0 ends immediately when there is no demand, 3 ends after 3 seconds of no demand, and max is the maximum duration of the cycle. Skip is only valid if there isn't a crosswalk. The most difficult part seems to be calculating elapsed time with no demand.

**Reply by dabreegster** [reply]
Implementation of adaptive signals would be really welcome! There was previous work by another contributor, but it's unfinished: #295 and https://github.com/dabreegster/abstreet/pull/298. There's some details about some of the necessary steps there that should still be current. In case you haven't found it, the UI for changing durations of stages is at https://github.com/dabreegster/abstreet/blob/master/game/src/edit/traffic_signals/edits.rs. The map model representation of signals is at https://github.com/dabreegster/abstreet/blob/master/map_model/src/objects/traffic_signals.rs, and when people save their edits in the game, it winds up converting that to a slightly different format in https://github.com/dabreegster/seattle_traffic_signals/blob/master/src/lib.rs. And finally, the simulation happens in https://github.com/dabreegster/abstreet/blob/master/sim/src/mechanics/intersection.rs. If you have any trouble wading through all of the code, feel free to ask here or [join our Slack](https://join.slack.com/t/abstreet/shared_invite/zt-k7ybmqlu-ToIwDxc2evgYNwSSKXGCng).

> The most difficult part seems to be calculating elapsed time with no demand

IIUC, if `skip` is true, we'd just immediately cut to the next stage. If not, the duration would be `min`. This could be evaluated somewhere near https://github.com/dabreegster/abstreet/blob/0fc0b8db65cc5e8413932c96ed0646d6cc9d29eb/sim/src/mechanics/intersection.rs#L255 -- maybe the way some of this works will have to change.

**Reply by BruceBrown** [reply]
The UI was pretty straightforward. I've refined the variant to be min, delay, additional. Skip is simply having a min of 0. The algorithm for advancing is just is demand, extend time by delay until total > addition in which case the last extend is a remainder. To keep track of that I think I'll have to extend the state to include the cumulative time of the stage. I've already pulled down the seattle_traffic_signals and modified it locally. I'm also considering adding lagging protected green for a turn. That's a bit more complex due to crosswalks, as well as editing per turn. I need to think about it some more. I want to complete the adaptive first, then add lagging later.

**Reply by dabreegster** [reply]
> I think I'll have to extend the state to include the cumulative time of the stage.

That makes sense; `SignalState` has `stage_ends_at`, you can probably stick the start time in there too and always subtract from `now`.

> I want to complete the adaptive first, then add lagging later

Sounds good! The adaptive stuff sounds like a great self-contained PR first.
